### ISSUE TITLE
Refactor Linux test and deploy Travis pipeline

### ISSUE THREAD

**Issue by ericzolf** [rdiff-backup/rdiff-backup#221]
@ardovm Where do you find all this free time to work on this ?

Many of the changes you made could/should be part of the makefile. Had recommend something similar:

**Solution1**
Makefile
```
bdist_wheel:
    docker run --rm -e PLAT=$PLAT -v `pwd`:/io $DOCKER_IMAGE $PRE_CMD /io/.travis/build_wheels.sh

```

.travis.yml
```
env:
  - MAKE_STEP=test-static RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=build RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=test-runtime-base RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=test-runtime-root RUN_COMMAND= SUDO=root  # see below
  - MAKE_STEP=test-runtime-slow RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=bdist_wheel DOCKER_IMAGE=quay.io/pypa/manylinux1_x86_64 PLAT=manylinux1_x86_64
  - MAKE_STEP=bdist_wheel DOCKER_IMAGE=quay.io/pypa/manylinux1_i686 PRE_CMD=linux32 PLAT=manylinux1_i686
  - ...

```

**Solution2**
Or something like this:

Makefile
```
bdist_wheel:
    docker run --rm -e PLAT=manylinux1_x86_64 -v `pwd`:/io quay.io/pypa/manylinux1_x86_64 /io/.travis/build_wheels.sh
    docker run --rm -e PLAT=manylinux1_i686 -v `pwd`:/io quay.io/pypa/manylinux1_i686 /io/.travis/build_wheels.sh
    ...

```

.travis.yml
```
env:
  - MAKE_STEP=test-static RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=build RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=test-runtime-base RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=test-runtime-root RUN_COMMAND= SUDO=root  # see below
  - MAKE_STEP=test-runtime-slow RUN_COMMAND= SUDO=sudo
  - MAKE_STEP=bdist_wheel
  - ...

``

I prefer the last one since it would make it very easy to generate wheel with a single command line on almost any linux environment with docker.

_Originally posted by @ikus060 in https://github.com/rdiff-backup/rdiff-backup/pull/219#issuecomment-569926459_

**Comment by ardovm** [comment]
@ikus060 I personally like solution 1 better, because it looks more coherent with the rule: «Makefile targets shall correspond to setup.py actions».

On the other hand, the exact solution you proposed, i.e.:
```Makefile
bdist_wheel:
    docker run --rm -e PLAT=manylinux1_x86_64 -v `pwd`:/io quay.io/pypa/manylinux1_x86_64 /io/.travis/build_wheels.sh
    docker run --rm -e PLAT=manylinux1_i686 -v `pwd`:/io quay.io/pypa/manylinux1_i686 /io/.travis/build_wheels.sh
```
would _require_ Docker, whereas the other Makefile targets use the `RUN_COMMAND` variable to make it optional.

I don't know Docker (the lines in my proposed .travis.yml were copied and pasted from other sources), therefore I am not sure what is the best way to go here.

Maybe adding another Makefile variable, such as `WHEEL_RUN_COMMAND` that invokes Docker as you suggested, and can be overridden to an empty string?

**Comment by ericzolf** [comment]
OK, the deployment works, even if we have many drafts, as @ardovm warned us.
[job_1302-Releases__302_267_rdiff-backup_rdiff-backup.pdf](https://github.com/rdiff-backup/rdiff-backup/files/4018500/job_1302-Releases__302_267_rdiff-backup_rdiff-backup.pdf)

First remark is that manylinux1 is released two times, whereas 2010 and 2014 are only released once. Not sure if it's due to our pipe or if Travis/GitHub are mixing up things when doing draft releases...

**Comment by ericzolf** [comment]
I realise that the duplicated manylinux1 releases don't have the same size, but that the manylinux1 packages released with the manylinux2010 packages have the same size as those same manylinux2010 packages. That is highly dubious. @ardovm can you have a look at this please?

And while we're at it, I would skip altogether the manylinux1 build:

1. it is based on CentOS 5 which is out of support hence a security hazard
2. as auditwheel potentially links libraries from the platform, we carry the same security hazard
3. there is no standard Python3 package for CentOS5 (not even in EPEL)

References:
* https://github.com/pypa/manylinux - "PEP 513 defined manylinux1_x86_64 and manylinux1_i686 platform tags and the wheels were built on Centos5. Centos5 reached End of Life (EOL) on March 31st, 2017 and thus PEP 571 was proposed."
**Side note:** I don't understand why manylinux2014 is offered but not referenced here...
* https://ftp-stud.hs-esslingen.de/pub/Mirrors/archive.fedoraproject.org/epel/5/SRPMS/ - no "python3" in there, only python 2.6

**Comment by ardovm** [comment]
@ericzolf I found out that the https://github.com/pypa/manylinux project has three branches:

- master: builds manylinux2010
- manylinux1: builds manylinux1
- manylinux2014: builds manylinux2014.

The [README for manylinux2014 ](https://github.com/pypa/manylinux/tree/manylinux2014) explains that it is based on PEP 599, CentOS 7. In practice, they are ready for when CentOS 6 (and therefore manylinux2010) will reach EOL, i.e. November 2020.

Thank you for your analysis of Python 3 packages. I agree that we could scrap manylinux1. I will start from this, and then look into the duplicate wheels problem, if it still appears.

**Comment by ardovm** [comment]
Confirmed: manylinux2010 wheels are also "duplicated" by `auditwheel repair` as manylinux1.

Excerpt from the build log:
```
+ auditwheel repair wheelhouse/rdiff_backup-0.1.dev53+gd2c83bc-cp36-cp36m-linux_x86_64.whl --plat manylinux2010_x86_64 -w /io/wheelhouse/
INFO:auditwheel.main_repair:Repairing rdiff_backup-0.1.dev53+gd2c83bc-cp36-cp36m-linux_x86_64.whl
INFO:auditwheel.wheeltools:Previous filename tags: linux_x86_64
INFO:auditwheel.wheeltools:New filename tags: manylinux2010_x86_64
INFO:auditwheel.wheeltools:Previous WHEEL info tags: cp36-cp36m-linux_x86_64
INFO:auditwheel.wheeltools:New WHEEL info tags: cp36-cp36m-manylinux2010_x86_64
INFO:auditwheel.main_repair:Wheel is eligible for a higher priority tag. You requested manylinux2010_x86_64 but I have found this wheel is eligible for manylinux1_x86_64.
INFO:auditwheel.wheeltools:Previous filename tags: linux_x86_64
INFO:auditwheel.wheeltools:New filename tags: manylinux1_x86_64
INFO:auditwheel.wheeltools:Previous WHEEL info tags: cp36-cp36m-linux_x86_64
INFO:auditwheel.wheeltools:New WHEEL info tags: cp36-cp36m-manylinux1_x86_64
INFO:auditwheel.main_repair:
Fixed-up wheel written to /io/wheelhouse/rdiff_backup-0.1.dev53+gd2c83bc-cp36-cp36m-manylinux1_x86_64.whl
```

I will open a PR to scrap manylinux1 from `.travis.yml`

**Comment by ottok** [comment]
Is the Travis-CI deploy responsible for making automatic draft releases?
![image](https://user-images.githubusercontent.com/668724/71901464-ae5e5900-3168-11ea-91cd-b2c2aad93035.png)

**Comment by ardovm** [comment]
> Is the Travis-CI deploy responsible for making automatic draft releases?

Yes ([see here for your reference](https://docs.travis-ci.com/user/deployment/releases/#draft-releases-with-draft-true)).

**Comment by ericzolf** [comment]
Now that the different changes to travis have been approved, I'm reviewing what's happening and starting the discussion to improve (hopefully) our current release pipeline:

1. the files in the Windows zip-file are _not_ in a sub-directory called something like `rdiff-backup-$version` (this was the case in the past).
2. why are the wheels created in a sub-dir named `wheelhouse`? I'd prefer them to be in `dist` like all other assets, this would be more consistent.
3. we are missing an `sdist` target in Makefile/.travis.yml so that we can release source code to packagers and to PyPI for the platforms without wheel (FreeBSD, MacOS, etc).
4. based on points 2 and 3, we could extend the wheel `deploy` section to also deploy sdist packages (also, later, to PyPI as well).
5. we don't need to "pip3-install" auditwheel any more as it's only called from within a container.
6. I'd like to switch from the situation today, where draft releases are created only on `-dodeploy` and `master` branches to working more with tags with a specific naming convention, it would make it IMHO easier to delete draft releases as they would be tagged, and tags are comparatively easy to create and delete.
7. once the above is implemented and works, I would create a beta release.

I can do all the above, but I wanted to be sure we're all aligned on what we want to reach before I start.

**Comment by ardovm** [comment]
Thank you for the review and the roadmap! All good for me.

My answer to point 2, for what it's worth: I used `wheelhouse` because it was an easy copy & paste from the original example. Moreover, the `dist` directory could also contain other files than the wheels alone, and I found it a bit confusing.

**Comment by ericzolf** [comment]
Closed by #244
